### PR TITLE
Quadratic model examples and docstrings

### PR DESCRIPTION
**Pull request by JoelPasvolsky**

### CODE REVIEW

**Review by arcondello — APPROVED** [review]

File: dimod/quadratic/quadratic_model.py [hunk 1]
```diff
@@ -469,6 +494,14 @@ def energies(self, samples_like, dtype: Optional[DTypeLike] = None) -> np.ndarra
         Returns:
             Energies for the samples.
 
+        Examples:
+            >>> from dimod import QuadraticModel, Binary
+            >>> qm = QuadraticModel()
+            >>> qm.add_variables_from('BINARY', ['x', 'y'])
+            >>> qm.add_quadratic('x', 'y', -2)
+            >>> qm.energies(([[1, 0], [0, 0], [1, 1]], ['x', 'y']))
```
[the remark this answers is not in the document]
Comment (JoelPasvolsky) [reply dimod/quadratic/quadratic_model.py:502]: Done

File: dimod/quadratic/quadratic_model.py [hunk 2]
```diff
@@ -487,7 +520,15 @@ def energy(self, sample, dtype=None) -> Bias:
...
         Returns:
-            The energy.
+            Energy for the sample.
+
+        Examples:
+            >>> from dimod import QuadraticModel, Binary
+            >>> qm = QuadraticModel()
+            >>> qm.add_variables_from('BINARY', ['x', 'y'])
+            >>> qm.add_quadratic('x', 'y', -2)
+            >>> qm.energy(([1, 1], ['x', 'y']))
```
[the remark this answers is not in the document]
Comment (JoelPasvolsky) [reply dimod/quadratic/quadratic_model.py:530]: Done

File: dimod/quadratic/quadratic_model.py [hunk 3]
```diff
@@ -497,7 +538,19 @@ def energy(self, sample, dtype=None) -> Bias:
...
+        """Construct a quadratic model from a binary quadratic model.
+
+        Args:
+            bqm: Binary quadratic model from which to create the quadratic model.
+
+        Returns:
+            Quadratic model.
+
+        Examples:
+            >>> from dimod import QuadraticModel, BinaryQuadraticModel, as_bqm
```
[the remark this answers is not in the document]
Comment (JoelPasvolsky) [reply dimod/quadratic/quadratic_model.py:550]: Changed to `bqm = BinaryQuadraticModel({'a': 0.1, 'b': 0.2}, {'ab': -1}, 'SPIN')`

File: dimod/quadratic/quadratic_model.py [hunk 4]
```diff
@@ -593,30 +647,83 @@ def is_linear(self) -> bool:
...
+        Returns:
+            Neighbors of the specified variable and their quadratic biases.
+
+        Examples:
+            >>> from dimod import QuadraticModel
+            >>> qm = QuadraticModel()
+            >>> qm.add_variables_from('BINARY', ['x', 'y', 'z'])
+            >>> qm.add_quadratic('x', 'y', -2)
+            >>> qm.add_quadratic('x', 'z', 2)
+            >>> [variable for variable in qm.iter_neighborhood('x')]
```
[the remark this answers is not in the document]
Comment (JoelPasvolsky) [reply dimod/quadratic/quadratic_model.py:664]: Changed to `list(qm.iter_neighborhood('x'))`

File: dimod/quadratic/quadratic_model.py [hunk 5]
```diff
@@ -593,30 +647,83 @@ def is_linear(self) -> bool:
...
+        Returns:
+            Interactions of the quadratic model and their biases.
+
+        Examples:
+            >>> from dimod import QuadraticModel
+            >>> qm = QuadraticModel()
+            >>> qm.add_variables_from('BINARY', ['x', 'y', 'z'])
+            >>> qm.add_quadratic('x', 'y', -2)
+            >>> qm.add_quadratic('x', 'z', 2)
+            >>> [variable for variable in qm.iter_quadratic()]
```
[the remark this answers is not in the document]
Comment (JoelPasvolsky) [reply dimod/quadratic/quadratic_model.py:683]: Ditto

File: dimod/quadratic/quadratic_model.py [hunk 6]
```diff
@@ -625,6 +732,31 @@ def reduce_neighborhood(self, v: Variable, function: Callable,
                             initializer: Optional[Bias] = None) -> Any:
         """Apply function of two arguments cumulatively to the quadratic biases
         associated with a single variable.
+
+        Args:
+
```
[the remark this answers is not in the document]
Comment (JoelPasvolsky) [reply dimod/quadratic/quadratic_model.py:737]: Deleted

File: dimod/quadratic/quadratic_model.py [hunk 7]
```diff
@@ -633,12 +765,51 @@ def reduce_quadratic(self, function: Callable,
...
+                mapped to the new labels.
+
+        Returns:
+            The original or new quadratic model with updated variable labels.
+
+        Examples:
+            >>> from dimod import QuadraticModel, BinaryQuadraticModel, generators
+            >>> bqm = generators.ran_r(1, 5)
+            >>> qm = QuadraticModel.from_bqm(bqm)
+            >>> qm_new = qm.relabel_variables({0: 'a', 1: 'b', 2: 'c'}, inplace=False)
```
[the remark this answers is not in the document]
Comment (JoelPasvolsky) [reply dimod/quadratic/quadratic_model.py:810]: Added `>>> qm_new.variables`

File: dimod/quadratic/quadratic_model.py [hunk 8]
```diff
@@ -593,30 +647,83 @@ def is_linear(self) -> bool:
...
+                the linear biases or used as the default if no linear biases are
+                set in the quadratic model.
+
+        Returns:
+            Result of applying the specified function to the linear biases.
+
+        Examples:
+            >>> from operator import add
+            >>> from dimod import QuadraticModel, BinaryQuadraticModel, generators
+            >>> bqm = generators.randint(5, 'BINARY', low=-5, high=5)
```
[the remark this answers is not in the document]
Comment (JoelPasvolsky) [reply dimod/quadratic/quadratic_model.py:720]: Updated to 
```
>>> from operator import add
            >>> from dimod import QuadraticModel
            >>> qm = QuadraticModel({'x': 0.5, 's': 1, 'i': 2}, 
            ...                     {('x', 'i'): 2}, 0.0, 
            ...                     {'x': 'BINARY', 's': 'SPIN', 'i': 'INTEGER'})
            >>> qm.reduce_linear(add) 
            3.5
```

File: dimod/quadratic/quadratic_model.py [hunk 9]
```diff
@@ -633,12 +765,51 @@ def reduce_quadratic(self, function: Callable,
...
+            initializer: Prefixed in the calculation to the iterable containing
+                the quadratic biases or used as the default if no quadratic biases
+                are set in the quadratic model.
+
+        Returns:
+            Result of applying the specified function to the quadratic biases.
+
+        Examples:
+            >>> from dimod import QuadraticModel, BinaryQuadraticModel, generators
+            >>> bqm = generators.randint(5, 'BINARY', low=-5, high=5)
```
[the remark this answers is not in the document]
Comment (JoelPasvolsky) [reply dimod/quadratic/quadratic_model.py:781]:
```
>>> qm = QuadraticModel({'x': 0.5, 's': 1, 'i': 2},
 ...                     {('x', 'i'): 2, ('s', 'i'): 3}, 0.0,
 ...                     {'x': 'BINARY', 's': 'SPIN', 'i': 'INTEGER'})
```

File: dimod/quadratic/quadratic_model.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (arcondello) [reply dimod/quadratic/quadratic_model.py:720]: If you think that's a bit unwieldy, you could construct it symbolically or with `add_linear` :shrug:

[the remark this answers is not in the document]
Comment (JoelPasvolsky) [reply dimod/quadratic/quadratic_model.py:720]: I think it works well:
image

File: dimod/quadratic/quadratic_model.py [hunk 10]
```diff
@@ -625,6 +732,31 @@ def reduce_neighborhood(self, v: Variable, function: Callable,
...
+                the quadratic biases or used as the default if variable ``v`` has
+                no quadratic biases.
+
+        Returns:
+            Result of applying the specified function to the specified variable's
+            quadratic biases.
+
+        Examples:
+            >>> from dimod import QuadraticModel, BinaryQuadraticModel, generators
+            >>> bqm = generators.randint(5, 'BINARY', low=-5, high=5)
```
[the remark this answers is not in the document]
Comment (JoelPasvolsky) [reply dimod/quadratic/quadratic_model.py:753]: Ditto

**Review by arcondello — APPROVED** [review]

**Pull request merged**